### PR TITLE
support unordered tlog message from the start

### LINKED ISSUE #166: tlog - nbd : tlog might expect wrong first sequence
After #161 merged, tlog might expect wrong first sequence if these cases happen:
- first message **received succesfully**  is not the first message **sent** by nbdserver.
- nbdserver restarted and reset the sequence to zero.

We can add the valid sequence as part of initial connection message (see #159)

### PR DESCRIPTION
**Pull request by GlenDC**
fixes #166

What changed:

+ TlogClient now sends `firstSequence` as part of `ClientVerAck` object. This `firstSequence` gets used as the first sequence of the tlog sequence for that vdisk within the tlog server. This allows us to support unordered tlog messages right from the start;
+ TlogServer now only accepts 1 connection per active vdisk. Thus when the server receives a VerAck, it tries to create a vdisk (using the vdiskmanager) with the vdiskID given by the client. If that vdisk already exists, the connection will not be established and the returned status will be an error status. If it doesn't exist yet, it will be created, and the connection gets successfully established. When the connection drops, the vdisk will be deleted again from the vdiskManager;

### CODE REVIEW

**Comment by iwanbk** [comment]
> TlogServer now only accepts 1 connection per active vdisk

How about the plan to support multiple connections?

File: tlog/tlogserver/server/vdisk.go [hunk 1]
```diff
@@ -75,25 +76,34 @@ func newVdiskManager(blockSize, flushSize int) *vdiskManager {
...
+	vt.vdisks[vdiskID] = vd
 	return vd, nil
 }
 
+// delete a vdisk (if possible)
+func (vt *vdiskManager) delete(vdiskID string) {
+	vt.lock.Lock()
+	defer vt.lock.Unlock()
+
+	delete(vt.vdisks, vdiskID)
```
Comment (iwanbk) [inline tlog/tlogserver/server/vdisk.go]: we also need to stop the goroutine if we want to delete it.(which i think is not necessary to fix this issue)

**Comment by GlenDC** [comment]
> How about the plan to support multiple connections?

.... That's not possible with this PR in place...

But @iwanbk, I did specify this new behavior in our telegram chat. 

If we do want multiple connections, I'll need to adapt this PR.

Let me know, but this will be for tomorrow, as I don't have much more time left to work.

File: tlog/tlogserver/server/vdisk.go (hunk 1 above)
Comment (GlenDC) [inline tlog/tlogserver/server/vdisk.go]: Already resolved, refresh.

**Comment by GlenDC** [comment]
If we do want to support multiple connections, than we do want to allow multiple connections on a single vdisk... however! In that case we do want to perhaps add some kind of reference counting within the vdisk manager, so that it can clean up automatically all goroutines once all connections for a vdisk are cleaned up. Because the old code simply leaves all goroutines open, which is not desired for long-running servers.

**Comment by iwanbk** [comment]
> But @iwanbk, I did specify this new behavior in our telegram chat.

Well, only behavior to reject newly connected client 

> If we do want multiple connections, I'll need to adapt this PR.

Yep, we need to close #128 for this PR to be accepted. Let me know if #128 is a bad idea that we need to close it

> If we do want to support multiple connections, than we do want to allow multiple connections on a single vdisk... 

It already does. vdisk only accept new tlog message via a single `inputChan` channel.

> however! In that case we do want to perhaps add some kind of reference counting within the vdisk manager, so that it can clean up automatically all goroutines once all connections for a vdisk are cleaned up. Because the old code simply leaves all goroutines open, which is not desired for long-running servers..


While you're right about stopping the vdisk object, it is not needed for this issue.

**Comment by GlenDC** [comment]
> While you're right about stopping the vdisk object, it is not needed for this issue.

Sure, but it also cannot wait for `1.1.0-alphe-4` I think...

>  Let me know if #128 is a bad idea that we need to close it

No, it's fine. Multiple connections does seem like a good enough idea, and your benchmarks do look good enough to support this more complex behavior.

**Comment by iwanbk** [comment]
> Sure, but it also cannot wait for 1.1.0-alpha-4 I think...

Yep, just saying :). 

It is also better to confirm the goal of the milestone, because AFAIK the tlogserver in this milestone only supposed to handle a single VM.

Cleaning goroutines also not as simple as stopping it, we also need to flush the unflushed blocks. And of course handle the case where we have missing blocks sequence.

**Comment by GlenDC** [comment]
> Cleaning goroutines also not as simple as stopping it, we also need to flush the unflushed blocks. And of course handle the case where we have missing blocks sequence.

That makes sense, can you add all this information to https://github.com/g8os/blockstor/issues/167 @iwanbk? Seems like we can fix that together with improving the general design of it all.

Will simply change the PR of this code then to not check for already connecting vdisks.

**Comment by GlenDC** [comment]
@iwanbk:

+ I rebased master into this PR;
+ Modified this PR's commit to allow multiple connections per vdisk, thus not validating if a vdisk doesn't exist yet;

In other words. I fixed your only objection, thus I think this is ready to be merged.

Please review and merge.

**Comment by iwanbk** [comment]
thanks @GlenDC. As a side note, can we rename `ClientVerAck` and 'ServerVerAck` to something else that more aligned with the content? Maybe handshakeReq and handshakeResp

I didn't ask for change in this PR because i already accept this naming in previous PR.

I already confused when it also contain vdiskID, it become even more confusing with `firstSequence` field addition.

**Comment by GlenDC** [comment]
> As a side note, can we rename ClientVerAck and 'ServerVerAck` to something else that more aligned with the content? Maybe handshakeReq and handshakeResp

Ok! I'll think about it and open a PR aimed at you.

**Pull request merged**